src/hotspot/os/linux/os_linux.cpp

```diff
@@ -5869,7 +5932,11 @@ static int checkpoint_restore() {
     return JVM_CHECKPOINT_ERROR;
   }
 
-  return info.si_int;
+  if (0 < info.si_int) {
+    *argp = get_new_args(info.si_int);
```

Comment:
Idea is exactly to avoid data corruption. For criu, it's a PID of the restoring process, that is, the PID that would java process receive with the usual start (without checkpoint/restore). The restoring process exists for the whole life of the restored java instance, so it is impossible for another restoring process to obtain the same ID. 

For pauseengine, there is a slight change of ID reuse, as restoring process may exit earlier than the target VM pick up the data. But since chances are small, and pauseengine is mostly for debugging, I don't think it is a big problem. To mitigate, another ID generation scheme could be used, e.g. the ID could be the PID of the target VM.